### ISSUE TITLE
How to access query parameters?

### ISSUE THREAD

**Issue by PandaWhisperer** [pandastrike/pbx#19]
Let's say I have a resource, and I have an action to list all instances of that resource:

``` coffeescript
# api.coffee
builder.define "widgets"
.get as: "query"

# handlers.coffee
widgets:
  query: async ({respond}) ->
     # obtain list of widgets
     respond 200, widgets
```

But I don't always want to list all widgets, I want to be able to filter them using query parameters. How do I do that?

I noticed that the `match` object seems to have a `query` key, so I tried that:

``` coffeescript
# handlers.coffee
widgets:
  query: async ({respond, match: {query}}) ->
     console.log query
     # obtain list of widgets
     respond 200, widgets
```

But when I do `curl localhost:1337/widgets?foo=bar`, I get `404 Not found`. Without the query parameter, it works.

**Comment by PandaWhisperer** [comment]
The more I look at the code, the more it appears to be a bug.

First, on [classifier.coffee#L41](https://github.com/pandastrike/pbx/blob/master/src/classifier.coffee#L41), the URL is parsed and pathname and query (as a hash) extracted. So far so good.

``` coffeescript
{pathname, query} = (url.parse request.url, true)
```

Then, on [classifier.coffee#L45](https://github.com/pandastrike/pbx/blob/master/src/classifier.coffee#L45), `resource.query.validate` is called on the `query`, and if that returns `false`, the response on [line 58](https://github.com/pandastrike/pbx/blob/master/src/classifier.coffee#L58) is `throw errors.not_found()`.

Now, what is `resource.query.validate`? The answer is a few lines earlier, on [lines 31-34](https://github.com/pandastrike/pbx/blob/master/src/classifier.coffee#L31-L34):

``` coffeescript
    node.resource =
      name: rname
      actions: {}
      query: {validate: (validator query)}
```

`query` in this context is defined on [line 28](https://github.com/pandastrike/pbx/blob/master/src/classifier.coffee#L28) to come from `api.mappings`:

``` coffeescript
{path, template, query} = api.mappings[rname]
```

The builder currently does not define a `query` key anywhere in the API mappings, so we can assume `query` will be `undefined`.

Now the [`validator`](https://github.com/pandastrike/pbx/blob/master/src/classifier.coffee#L8-L21) method returns a function that either validates against a schema, or, if called with a falsey value (such as in this case), it returns a function that returns `true` if called with `undefined` or `null`, and otherwise, it returns `true` only if the parameter has no keys (i.e. it is an empty object). 

``` coffeescript
    (object) ->
      # handle null, undefined, or {}
      # to mean 'empty query string'
      unless object?
        true
      else
        Object.keys(object).length == 0
```

In other words, it returns `false` anytime it is called with a non-empty object (i.e. when query parameters were present in the request url. Therefore the test on [line 45](https://github.com/pandastrike/pbx/blob/master/src/classifier.coffee#L45) will fail, and the response is a 404 error.

Anything from here on out is pure speculation, but my guess is that support for query parameters was planned, including the ability to validate queries using JSON schema. Support in the context is already present, but the builder currently doesn't let us define the appropriate settings.

**Comment by dyoder** [comment]
You have to define the resource to take query parameters. There is a `query` property for a resource that defines the query schema. Unfortunately, I forgot to add a way to add that using the Builder. Closing in favor of #20.

**Comment by PandaWhisperer** [comment]
Indeed. I see there's even a test for query parameters. But yeah, no builder support. Looking forward to that.
